### PR TITLE
Support CUDA 12.2

### PR DESCRIPTION
**Pull request by jameslamb**
## Description

* switches to CUDA 12.2.2 for building conda packages and wheels
* adds new tests running against CUDA 12.2.2

### Notes for Reviewers

This is part of ongoing work to build and test packages against CUDA 12.2.2 across all of RAPIDS.

For more details see:

* https://github.com/rapidsai/build-planning/issues/7
* https://github.com/rapidsai/shared-workflows/pull/166

Planning a second round of PRs to revert these references back to a proper `branch-24.{nn}` release branch of `shared-workflows` once https://github.com/rapidsai/shared-workflows/pull/166 is merged.

*(created with `rapids-reviser`)*

### CODE REVIEW

**Comment by jameslamb** [comment]
conda builds and tests on CUDA 12.0.1 and 12.2.2 are segfaulting ([build link](https://github.com/rapidsai/ucxx/actions/runs/7506630964/job/20438790429?pr=161)).

I don't see similar errors on recent builds on `branch-0.36` or other PRs.

**Comment by pentschev** [comment]
> conda builds and tests on CUDA 12.0.1 and 12.2.2 are segfaulting ([build link](https://github.com/rapidsai/ucxx/actions/runs/7506630964/job/20438790429?pr=161)).
> 
> I don't see similar errors on recent builds on `branch-0.36` or other PRs.

Those are indeed flaky tests, they don't always happen in same form. Let's first try rerunning them, based on the log for 12.2 almost all tests passed, the one that failed is one of the very last ones and likely to pass soon.

**Comment by jameslamb** [comment]
Ok thank you!

**Comment by jakirkham** [comment]
Thanks James & Peter! 🙏

Looks like that cleared things up and builds now pass

Though am going to mark this "do not merge" for the moment while we work through issues in the other PRs

**Comment by jakirkham** [comment]
Looks like there is a conflict here

It is a generated file. So can just regenerate it with [`rapids-dependency-file-generator`]( https://github.com/rapidsai/dependency-file-generator )

**Comment by jameslamb** [comment]
> Looks like there is a conflict here

fixed in https://github.com/rapidsai/ucxx/pull/161/commits/3dbdb61fac8d3b99b7b91ceb74c72f136be342cf

**Comment by jameslamb** [comment]
I see a new test failure on the v11.8.0 conda tests.

There are quite a few stacktraces and things that look like errors in the logs, but only a single unit test case failure.

I think these snippets summarize it well:

```text
2024-01-25 20:38:01,988 - distributed.worker - ERROR - Failed to communicate with scheduler during heartbeat.
Traceback (most recent call last):
  File "/tmp/distributed/distributed/core.py", line 1564, in _connect
    comm = await connect(
  File "/tmp/distributed/distributed/comm/core.py", line 342, in connect
    comm = await wait_for(
  File "/tmp/distributed/distributed/utils.py", line 1940, in wait_for
    return await asyncio.wait_for(fut, timeout)
  File "/opt/conda/envs/test/lib/python3.10/asyncio/tasks.py", line 432, in wait_for
    await waiter
asyncio.exceptions.CancelledError
...
distributed.comm.core.CommClosedError: ConnectionPool closing.
asyncio.exceptions.InvalidStateError: invalid state
asyncio.exceptions.InvalidStateError: invalid state
...
_____________________ ERROR at teardown of test_transpose ______________________

    @pytest.fixture(scope="function")
    def ucxx_loop():
...
        with check_thread_leak():
            yield loop
>           ucxx.reset()
...
>               raise UCXError(msg)
E               ucxx.UCXXError: Trying to reset UCX but not all Endpoints and/or Listeners are closed(). The following objects are still referencing ApplicationContext: 
E                 <frame at 0x7fdfe400e4c0, file '/opt/conda/envs/test/lib/python3.10/site-packages/ucxx/_lib_async/listener.py', line 155, code _listener_handler_coroutine>
E                 <frame at 0x55ddca529910, file '/opt/conda/envs/test/lib/python3.10/site-packages/ucxx/_lib_async/listener.py', line 155, code _listener_handler_coroutine>
...
/opt/conda/envs/test/lib/python3.10/site-packages/ucxx/core.py:111: UCXXError
ERROR python/distributed-ucxx/distributed_ucxx/tests/test_ucxx.py::test_transpose - ucxx.UCXXError: Trying to reset UCX but not all Endpoints and/or Listeners are closed(). The following objects are still referencing ApplicationContext: 
  <frame at 0x7fdfe400e4c0, file '/opt/conda/envs/test/lib/python3.10/site-packages/ucxx/_lib_async/listener.py', line 155, code _listener_handler_coroutine>
  <frame at 0x55ddca529910, file '/opt/conda/envs/test/lib/python3.10/site-packages/ucxx/_lib_async/listener.py', line 155, code _listener_handler_coroutine>
```

([build link](https://github.com/rapidsai/ucxx/actions/runs/7659213469/job/20874284961?pr=161))

**NOTE:** as of #174, `rapids-dask-dependency` is getting pulled in, so the tests are getting the latest `main` branch of `dask` and `distributed`, not releases. Maybe this error lies in some change between the latest `distributed` / `dask` releases and current `main` branch of those projects.

**Comment by jakirkham** [comment]
Vyas restarted CI as we believe this is an unrelated flaky test

**Review by vyasr — CHANGES_REQUESTED** [review]
I'm guessing it's because this PR was started before #145, but it needs updating for the wheel jobs. The conda outputs look sensible.

I was going to ask some questions about the necessity of the CUDA dependency in ucx-py because it's pure Python, but I realize those are less relevant here since there is an actual C++ dependency in ucxx since it is not a pure Python package and we do need to compile against CUDA.

File: .github/workflows/build.yaml
Comment (AyodeAwe) [inline .github/workflows/build.yaml:1]: The description in this PR mentions "switching to 12.2 for build conda packages and wheels" - should the branch reference [here](https://github.com/rapidsai/ucxx/blob/9fc63e078e7073bf9bdab8f4ffadb074178c3502/.github/workflows/build.yaml#L62) (and in other places) be changed as well?

If left as is, it doesn't look like the wheel for `ucxx` will be building against CUDA 12.2

Comment (bdice) [reply .github/workflows/build.yaml:1]: These wheels were newly added since this PR was created. I can help fix this.

Comment (bdice) [reply .github/workflows/build.yaml:1]: Done in e75732a3a668af249b0de72a2af715d84ae9959d.

**Review by AyodeAwe — APPROVED** [review]

**Review by jakirkham — APPROVED** [review]

**Review by vyasr — APPROVED** [review]

**Pull request merged**